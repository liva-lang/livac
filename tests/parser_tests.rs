use insta::assert_snapshot;
use livac::{lexer::tokenize, parser::parse};
use std::fs;

/// Test helper para casos correctos del parser
fn test_parser_ok(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/ok_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: ok_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let ast = parse(tokens, &source).unwrap();
    
    // Convertir AST a JSON para snapshot
    let json = serde_json::to_string_pretty(&ast).unwrap();
    assert_snapshot!(format!("ok_{}.ast", test_name), json);
}

/// Test helper para casos de error del parser
fn test_parser_err(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/err_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: err_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let result = parse(tokens, &source);
    assert!(result.is_err(), "Expected parser error for test: {}", test_name);
    
    let error_msg = result.unwrap_err().to_string();
    assert_snapshot!(format!("err_{}.diag", test_name), error_msg);
}

#[test]
fn test_functions_oneliner() {
    test_parser_ok("functions_oneliner");
}

#[test]
fn test_functions_block() {
    test_parser_ok("functions_block");
}

#[test]
fn test_classes() {
    test_parser_ok("classes");
}

#[test]
fn test_control_flow() {
    test_parser_ok("control_flow");
}

#[test]
fn test_expressions() {
    test_parser_ok("expressions");
}

#[test]
fn test_imports() {
    test_parser_ok("imports");
}

#[test]
fn test_unclosed_paren() {
    test_parser_err("unclosed_paren");
}

#[test]
fn test_unclosed_brace() {
    test_parser_err("unclosed_brace");
}

#[test]
fn test_case_without_switch() {
    test_parser_err("case_without_switch");
}

#[test]
fn test_return_outside_function() {
    test_parser_err("return_outside_function");
}

#[test]
fn test_duplicate_default() {
    test_parser_err("duplicate_default");
}
