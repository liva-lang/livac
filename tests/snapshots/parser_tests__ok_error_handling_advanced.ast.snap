---
source: tests/parser_tests.rs
expression: json
---
{
  "items": [
    {
      "Function": {
        "name": "readFile",
        "type_params": [],
        "params": [
          {
            "name": "filename",
            "type_ref": {
              "Simple": "string"
            },
            "default": null
          }
        ],
        "return_type": {
          "Simple": "string"
        },
        "body": {
          "stmts": [
            {
              "TryCatch": {
                "try_block": {
                  "stmts": [
                    {
                      "If": {
                        "condition": {
                          "Binary": {
                            "op": "Eq",
                            "left": {
                              "Identifier": "filename"
                            },
                            "right": {
                              "Literal": {
                                "String": "nonexistent.txt"
                              }
                            }
                          }
                        },
                        "then_branch": {
                          "Block": {
                            "stmts": [
                              {
                                "Throw": {
                                  "expr": {
                                    "Literal": {
                                      "String": "File not found"
                                    }
                                  }
                                }
                              }
                            ]
                          }
                        },
                        "else_branch": null
                      }
                    },
                    {
                      "Return": {
                        "expr": {
                          "Literal": {
                            "String": "file content"
                          }
                        }
                      }
                    }
                  ]
                },
                "catch_var": "e",
                "catch_block": {
                  "stmts": [
                    {
                      "Throw": {
                        "expr": {
                          "StringTemplate": {
                            "parts": [
                              {
                                "Text": "Failed to read file: "
                              },
                              {
                                "Expr": {
                                  "Identifier": "e"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  ]
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "safeDivide",
        "type_params": [],
        "params": [
          {
            "name": "a",
            "type_ref": {
              "Simple": "number"
            },
            "default": null
          },
          {
            "name": "b",
            "type_ref": {
              "Simple": "number"
            },
            "default": null
          }
        ],
        "return_type": {
          "Generic": {
            "base": "Result",
            "args": [
              {
                "Simple": "number"
              },
              {
                "Simple": "string"
              }
            ]
          }
        },
        "body": {
          "stmts": [
            {
              "If": {
                "condition": {
                  "Binary": {
                    "op": "Eq",
                    "left": {
                      "Identifier": "b"
                    },
                    "right": {
                      "Literal": {
                        "Int": 0
                      }
                    }
                  }
                },
                "then_branch": {
                  "Block": {
                    "stmts": [
                      {
                        "Return": {
                          "expr": {
                            "Call": {
                              "callee": {
                                "Identifier": "Err"
                              },
                              "args": [
                                {
                                  "Literal": {
                                    "String": "Division by zero"
                                  }
                                }
                              ],
                              "exec_policy": "normal",
                              "type_args": []
                            }
                          }
                        }
                      }
                    ]
                  }
                },
                "else_branch": null
              }
            },
            {
              "Return": {
                "expr": {
                  "Call": {
                    "callee": {
                      "Identifier": "Ok"
                    },
                    "args": [
                      {
                        "Binary": {
                          "op": "Div",
                          "left": {
                            "Identifier": "a"
                          },
                          "right": {
                            "Identifier": "b"
                          }
                        }
                      }
                    ],
                    "exec_policy": "normal",
                    "type_args": []
                  }
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "processDivision",
        "type_params": [],
        "params": [
          {
            "name": "a",
            "type_ref": {
              "Simple": "number"
            },
            "default": null
          },
          {
            "name": "b",
            "type_ref": {
              "Simple": "number"
            },
            "default": null
          }
        ],
        "return_type": {
          "Simple": "string"
        },
        "body": {
          "stmts": [
            {
              "TryCatch": {
                "try_block": {
                  "stmts": [
                    {
                      "VarDecl": {
                        "bindings": [
                          {
                            "name": "result",
                            "type_ref": null
                          }
                        ],
                        "init": {
                          "Call": {
                            "callee": {
                              "Identifier": "safeDivide"
                            },
                            "args": [
                              {
                                "Identifier": "a"
                              },
                              {
                                "Identifier": "b"
                              }
                            ],
                            "exec_policy": "normal",
                            "type_args": []
                          }
                        },
                        "is_fallible": false
                      }
                    },
                    {
                      "If": {
                        "condition": {
                          "MethodCall": {
                            "object": {
                              "Identifier": "result"
                            },
                            "method": "isOk",
                            "args": [],
                            "adapter": "seq",
                            "adapter_options": {}
                          }
                        },
                        "then_branch": {
                          "Block": {
                            "stmts": [
                              {
                                "Return": {
                                  "expr": {
                                    "StringTemplate": {
                                      "parts": [
                                        {
                                          "Text": "Success: "
                                        },
                                        {
                                          "Expr": {
                                            "MethodCall": {
                                              "object": {
                                                "Identifier": "result"
                                              },
                                              "method": "unwrap",
                                              "args": [],
                                              "adapter": "seq",
                                              "adapter_options": {}
                                            }
                                          }
                                        }
                                      ]
                                    }
                                  }
                                }
                              }
                            ]
                          }
                        },
                        "else_branch": {
                          "Block": {
                            "stmts": [
                              {
                                "Return": {
                                  "expr": {
                                    "StringTemplate": {
                                      "parts": [
                                        {
                                          "Text": "Error: "
                                        },
                                        {
                                          "Expr": {
                                            "MethodCall": {
                                              "object": {
                                                "Identifier": "result"
                                              },
                                              "method": "unwrapErr",
                                              "args": [],
                                              "adapter": "seq",
                                              "adapter_options": {}
                                            }
                                          }
                                        }
                                      ]
                                    }
                                  }
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  ]
                },
                "catch_var": "e",
                "catch_block": {
                  "stmts": [
                    {
                      "Return": {
                        "expr": {
                          "StringTemplate": {
                            "parts": [
                              {
                                "Text": "Unexpected error: "
                              },
                              {
                                "Expr": {
                                  "Identifier": "e"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  ]
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "complexErrorHandling",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "TryCatch": {
                "try_block": {
                  "stmts": [
                    {
                      "VarDecl": {
                        "bindings": [
                          {
                            "name": "data",
                            "type_ref": null
                          }
                        ],
                        "init": {
                          "Call": {
                            "callee": {
                              "Identifier": "readFile"
                            },
                            "args": [
                              {
                                "Literal": {
                                  "String": "nonexistent.txt"
                                }
                              }
                            ],
                            "exec_policy": "normal",
                            "type_args": []
                          }
                        },
                        "is_fallible": false
                      }
                    },
                    {
                      "TryCatch": {
                        "try_block": {
                          "stmts": [
                            {
                              "VarDecl": {
                                "bindings": [
                                  {
                                    "name": "result",
                                    "type_ref": null
                                  }
                                ],
                                "init": {
                                  "Call": {
                                    "callee": {
                                      "Identifier": "safeDivide"
                                    },
                                    "args": [
                                      {
                                        "Literal": {
                                          "Int": 10
                                        }
                                      },
                                      {
                                        "Literal": {
                                          "Int": 0
                                        }
                                      }
                                    ],
                                    "exec_policy": "normal",
                                    "type_args": []
                                  }
                                },
                                "is_fallible": false
                              }
                            },
                            {
                              "Expr": {
                                "expr": {
                                  "Call": {
                                    "callee": {
                                      "Identifier": "print"
                                    },
                                    "args": [
                                      {
                                        "StringTemplate": {
                                          "parts": [
                                            {
                                              "Text": "Result: "
                                            },
                                            {
                                              "Expr": {
                                                "Identifier": "result"
                                              }
                                            }
                                          ]
                                        }
                                      }
                                    ],
                                    "exec_policy": "normal",
                                    "type_args": []
                                  }
                                }
                              }
                            }
                          ]
                        },
                        "catch_var": "e",
                        "catch_block": {
                          "stmts": [
                            {
                              "Expr": {
                                "expr": {
                                  "Call": {
                                    "callee": {
                                      "Identifier": "print"
                                    },
                                    "args": [
                                      {
                                        "StringTemplate": {
                                          "parts": [
                                            {
                                              "Text": "Division error: "
                                            },
                                            {
                                              "Expr": {
                                                "Identifier": "e"
                                              }
                                            }
                                          ]
                                        }
                                      }
                                    ],
                                    "exec_policy": "normal",
                                    "type_args": []
                                  }
                                }
                              }
                            }
                          ]
                        }
                      }
                    }
                  ]
                },
                "catch_var": "e",
                "catch_block": {
                  "stmts": [
                    {
                      "Expr": {
                        "expr": {
                          "Call": {
                            "callee": {
                              "Identifier": "print"
                            },
                            "args": [
                              {
                                "StringTemplate": {
                                  "parts": [
                                    {
                                      "Text": "File error: "
                                    },
                                    {
                                      "Expr": {
                                        "Identifier": "e"
                                      }
                                    }
                                  ]
                                }
                              }
                            ],
                            "exec_policy": "normal",
                            "type_args": []
                          }
                        }
                      }
                    },
                    {
                      "Throw": {
                        "expr": {
                          "StringTemplate": {
                            "parts": [
                              {
                                "Text": "Critical error in complexErrorHandling: "
                              },
                              {
                                "Expr": {
                                  "Identifier": "e"
                                }
                              }
                            ]
                          }
                        }
                      }
                    }
                  ]
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "handleMultipleErrors",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "TryCatch": {
                "try_block": {
                  "stmts": [
                    {
                      "VarDecl": {
                        "bindings": [
                          {
                            "name": "result",
                            "type_ref": null
                          }
                        ],
                        "init": {
                          "Call": {
                            "callee": {
                              "Identifier": "parseAndDivide"
                            },
                            "args": [
                              {
                                "Literal": {
                                  "String": "10"
                                }
                              },
                              {
                                "Literal": {
                                  "String": "0"
                                }
                              }
                            ],
                            "exec_policy": "normal",
                            "type_args": []
                          }
                        },
                        "is_fallible": false
                      }
                    },
                    {
                      "Expr": {
                        "expr": {
                          "Call": {
                            "callee": {
                              "Identifier": "print"
                            },
                            "args": [
                              {
                                "StringTemplate": {
                                  "parts": [
                                    {
                                      "Text": "Result: "
                                    },
                                    {
                                      "Expr": {
                                        "Identifier": "result"
                                      }
                                    }
                                  ]
                                }
                              }
                            ],
                            "exec_policy": "normal",
                            "type_args": []
                          }
                        }
                      }
                    }
                  ]
                },
                "catch_var": "e",
                "catch_block": {
                  "stmts": [
                    {
                      "If": {
                        "condition": {
                          "MethodCall": {
                            "object": {
                              "Identifier": "e"
                            },
                            "method": "contains",
                            "args": [
                              {
                                "Literal": {
                                  "String": "parse"
                                }
                              }
                            ],
                            "adapter": "seq",
                            "adapter_options": {}
                          }
                        },
                        "then_branch": {
                          "Block": {
                            "stmts": [
                              {
                                "Expr": {
                                  "expr": {
                                    "Call": {
                                      "callee": {
                                        "Identifier": "print"
                                      },
                                      "args": [
                                        {
                                          "StringTemplate": {
                                            "parts": [
                                              {
                                                "Text": "Parse error: "
                                              },
                                              {
                                                "Expr": {
                                                  "Identifier": "e"
                                                }
                                              }
                                            ]
                                          }
                                        }
                                      ],
                                      "exec_policy": "normal",
                                      "type_args": []
                                    }
                                  }
                                }
                              }
                            ]
                          }
                        },
                        "else_branch": {
                          "Stmt": {
                            "If": {
                              "condition": {
                                "MethodCall": {
                                  "object": {
                                    "Identifier": "e"
                                  },
                                  "method": "contains",
                                  "args": [
                                    {
                                      "Literal": {
                                        "String": "division"
                                      }
                                    }
                                  ],
                                  "adapter": "seq",
                                  "adapter_options": {}
                                }
                              },
                              "then_branch": {
                                "Block": {
                                  "stmts": [
                                    {
                                      "Expr": {
                                        "expr": {
                                          "Call": {
                                            "callee": {
                                              "Identifier": "print"
                                            },
                                            "args": [
                                              {
                                                "StringTemplate": {
                                                  "parts": [
                                                    {
                                                      "Text": "Division error: "
                                                    },
                                                    {
                                                      "Expr": {
                                                        "Identifier": "e"
                                                      }
                                                    }
                                                  ]
                                                }
                                              }
                                            ],
                                            "exec_policy": "normal",
                                            "type_args": []
                                          }
                                        }
                                      }
                                    }
                                  ]
                                }
                              },
                              "else_branch": {
                                "Block": {
                                  "stmts": [
                                    {
                                      "Expr": {
                                        "expr": {
                                          "Call": {
                                            "callee": {
                                              "Identifier": "print"
                                            },
                                            "args": [
                                              {
                                                "StringTemplate": {
                                                  "parts": [
                                                    {
                                                      "Text": "Other error: "
                                                    },
                                                    {
                                                      "Expr": {
                                                        "Identifier": "e"
                                                      }
                                                    }
                                                  ]
                                                }
                                              }
                                            ],
                                            "exec_policy": "normal",
                                            "type_args": []
                                          }
                                        }
                                      }
                                    }
                                  ]
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  ]
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    },
    {
      "Function": {
        "name": "main",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "Expr": {
                "expr": {
                  "Call": {
                    "callee": {
                      "Identifier": "print"
                    },
                    "args": [
                      {
                        "Literal": {
                          "String": "Advanced error handling parsed successfully"
                        }
                      }
                    ],
                    "exec_policy": "normal",
                    "type_args": []
                  }
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    }
  ]
}
